feat: Implementation of seed finding in SYCL

czangela opened the pull request:
This is the implementation of the seed finding algorithm in SYCL, that I tried to present [in this talk](https://indico.cern.ch/event/948829/).

A few changes have been applied since, such as
 - switching to explicit memory management (with USM)
 - using nd_ranges to specify block size explicitly

Some time metrics on the outdated (?) test files of other seed finding unit tests.
(I will run tests on the new files as well, but according to @krasznaa, they don't provide a significant performance change, as we assumed previously.)

Architecture:
- CPU: IntelR CoreTM i9-9900K Processor (16M Cache, up to 5.00 GHz)
- NVIDIA GPU: GeForce RTX 2060 6GB GDDR6
(compiler: clang version 11.0.0 (built: 2020-08-24))

(By default, binaries are generated for OpenCL devices, so in order to run on NVIDIA GPU-s, the SYCL_POSSIBLE_TARGETS option should be set, for example to "nvptx64-nvidia-cuda-sycldevice".)

```bash
[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp1.txt -d GeForce -m
Running on: GeForce RTX 2060
read 85808 SP from file ../../seeds/sp1.txt
Preparation time: 0.688730
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):   0.606111    0.266637            2.273169
        Seeds found:      19759       19759          100.000000
---------------------------------------------------------------

[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp2.txt -d GeForce -m
Running on: GeForce RTX 2060
read 158051 SP from file ../../seeds/sp2.txt
Preparation time: 0.779685
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):   3.830942    1.117647            3.427686
        Seeds found:      80638       80637           99.600685
---------------------------------------------------------------

[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp5.txt -d GeForce -m
Running on: GeForce RTX 2060
read 342211 SP from file ../../seeds/sp5.txt
Preparation time: 1.006188
Analyzed 260 groups for CPU

Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s): 329.472830  165.669326            1.988738
        Seeds found:     358440      358439           97.728210
---------------------------------------------------------------

[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp10.txt -d GeForce -m -n 50
Running on: GeForce RTX 2060
read 749018 SP from file ../../seeds/sp10.txt
Preparation time: 1.524057
Analyzed 50 groups for CPU
Analyzed 50 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):  64.949634   31.722718            2.047417
        Seeds found:     128086      128086           99.077965
---------------------------------------------------------------
```

Same on Intel graphics card:
 - Intel GPU: Intel Gen9 HD Graphics NEO

```bash
[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp1.txt -m
Running on: Intel(R) Gen9 HD Graphics NEO
read 85808 SP from file ../../seeds/sp1.txt
Preparation time: 0.649866
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):   0.603116    0.684943            0.880534
        Seeds found:      19759       19759           99.994942
---------------------------------------------------------------

[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp2.txt -m
Running on: Intel(R) Gen9 HD Graphics NEO
read 158051 SP from file ../../seeds/sp2.txt
Preparation time: 0.738565
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):   3.965069    4.070089            0.974197
        Seeds found:      80638       80636           99.616806
---------------------------------------------------------------

[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp5.txt -m -n 20
Running on: Intel(R) Gen9 HD Graphics NEO
read 342211 SP from file ../../seeds/sp5.txt
Preparation time: 0.963241
Analyzed 20 groups for CPU
Analyzed 20 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):   7.103991   11.170573            0.635956
        Seeds found:      23257       23257           98.542374
---------------------------------------------------------------

[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp10.txt -m -n 10
Running on: Intel(R) Gen9 HD Graphics NEO
read 749018 SP from file ../../seeds/sp10.txt
Preparation time: 1.504190
Analyzed 10 groups for CPU
Analyzed 10 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):  24.690320   46.565883            0.530223
        Seeds found:      29169       29169           99.177208
---------------------------------------------------------------
```

Implementation details will be provided in the comments.

(+ more runtime metrics)

krasznaa commented on Plugins/CMakeLists.txt:
```diff
@@ -9,5 +9,6 @@ add_component_if(Legacy PluginLegacy ACTS_BUILD_PLUGIN_LEGACY)
+add_component_if(Sycl PluginSycl ACTS_BUILD_PLUGIN_SYCL)
 
-propagate_components_to_parent()
+propagate_components_to_parent()
```
Please put the newline back at the end of the file. :wink:

krasznaa commented on Plugins/Sycl/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+include( CMakeParseArguments )
+set( CMAKE_REQUIRED_QUIET TRUE )
+
+# set(SYCL_DIR ${CMAKE_CURRENT_SOURCE_DIR})
```
Apart from the `cmake_minimum_required(...)` line nothing before (and including) this line should be necessary for this plugin.

krasznaa commented on Plugins/Sycl/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+  src/Seeding/CreateSeedsForGroupSycl.cpp
+  src/Utilities/CreateQueueSycl.cpp
+  src/Utilities/ListSyclPlatforms.cpp
+  src/Utilities/DeviceSelector.cpp
```
This is just a personal preference, but I like to also list the header files that belong to a library/executable. Certain IDEs (most notably [Visual Studio](https://visualstudio.microsoft.com/) and [Xcode](https://developer.apple.com/xcode/)) set up much nicer projects with the header files listed as well.

czangela replied:
ok

krasznaa commented on Plugins/Sycl/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+  
+install(
+  DIRECTORY include/Acts
+  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
```
Newline...

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.hpp:
```diff
@@ -0,0 +1,81 @@
+/// creating the queue in a standalone translation unit.
+///
+/// @return A pointer to the queue.
+std::shared_ptr<cl::sycl::queue> createQueue(const std::string&);
```
As we discussed before, I'd prefer to have the definitions of both of these functions in their own header files.

czangela replied:
Oh yes, I remember now.

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.ipp:
```diff
@@ -0,0 +1,123 @@
+    deviceBottomSPs.insert(
+        deviceBottomSPs.end(),
+        deviceSpacePoint{SP->x(), SP->y(), SP->z(), SP->radius(),
+                         SP->varianceR(), SP->varianceZ()});
```
No need to use `insert(...)` here. `push_back(...)` should be just fine as well.

czangela replied:
Yes, that was legacy, and I forgot to change it. Thanks.

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.ipp:
```diff
@@ -0,0 +1,123 @@
+      m_config.seedFilter->getSeedFilterConfig().compatSeedWeight,
+      m_config.impactMax,
+      m_config.seedFilter->getSeedFilterConfig().compatSeedLimit,
+  };
```
Why don't you make this object a member of `Acts::Sycl::SeedFinder`? There is no need to create it from scratch on every call.

czangela replied:
I did not want to change the interface of the Seedfinder class too much.

krasznaa replied:
:confused: You don't have to. You don't even need to change the list of arguments received by the constructor.

Just created a variable called `Acts::Sycl::SeedFinder::m_deviceConfigData`, and set it up in the constructor of the class. With this exact payload.

czangela replied:
Ok, I'll change this in the next commit.

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/detail/Types.h:
```diff
@@ -0,0 +1,65 @@
+#include <bits/stdint-uintn.h>
+
+namespace Acts::Sycl::detail {
+struct deviceSpacePoint {
```
We will need input from the Acts developers, but I guess they took the convention from ATLAS of starting every class/struct name with a capital letter. (In ATLAS offline code we use this convention to more easily differentiate between "simple" functions and more complicated constructs like functors.)

msmk0 replied:
I would refer you to the [coding guidelines](https://acts.readthedocs.io/en/latest/codeguide.html), specifically the naming section.

Btw. I am not sure the `#include <bits/stdint-uintn.h>` is a standard include. Could you just use `#include <cstdint>` instead?

krasznaa replied:
I agree. I missed that include...

czangela replied:
Thanks

krasznaa replied:
The code still didn't change the capitalisation of these classes. :confused: The discussion should be kept open until the issue is resolved...

krasznaa commented on Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp:
```diff
@@ -0,0 +1,712 @@
+// Returns the smallest multiple of workGroupSize that is not smaller
+// than numThreads.
+uint32_t numWorkItems(uint32_t numThreads, uint32_t workGroupSize) {
+  auto q = numThreads / workGroupSize + (numThreads % workGroupSize != 0);
```
You can achieve the same with:

```c++
(numThreads + workGroupSize - 1) / workGroupSize
```

Which avoids using the `%` operator. (Which can be rather slow.)

krasznaa commented on Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp:
```diff
@@ -0,0 +1,712 @@
+              sizeof(detail::deviceSpacePoint) * (T));
+    q->memset(deviceCountBotDuplets, 0, M * sizeof(std::atomic_uint32_t));
+    q->memset(deviceCountTopDuplets, 0, M * sizeof(std::atomic_uint32_t));
+    q->wait();
```
I don't think this `wait()` is necessary here...

krasznaa commented on Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp:
```diff
@@ -0,0 +1,712 @@
+      edgesComb = sumBotTopCombined[M];
+
+      if (edgesBottom == 0 || edgesTop == 0)
+        return;
```
What happens with all the allocated device memory when we return here? Yepp, it leaks... :frowning:

czangela replied:
Omg, thanks.

krasznaa commented on Tests/UnitTests/Plugins/Sycl/Seeding/CommandLineArguments.h:
```diff
@@ -0,0 +1,21 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+#include <fstream>
```
You need an include for `<string>` as well.

krasznaa commented on Tests/UnitTests/Plugins/Sycl/Seeding/CommandLineArguments.h:
```diff
@@ -0,0 +1,21 @@
+  bool allgroup = false;
+  bool onlyGpu = false;
+  bool matches = false;
+  uint groups = 500;
```
I don't like this type... You should just use the C\+\+ standard `unsigned int` type instead.

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.hpp:
```diff
@@ -0,0 +1,81 @@
+ private:
+  Acts::SeedfinderConfig<external_spacepoint_t> m_config;
+  Acts::Sycl::DeviceExperimentCuts m_deviceCuts;
+  std::shared_ptr<cl::sycl::queue> m_queue;
```
:confused: I find it curious that [std::shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr) works correctly when a given compilation unit only sees a forward declaration of the type that it uses...

I think this is a pretty fragile/platform dependent setup. :frowning: As we discussed before, I think it would be much safer to handle the queue through a wrapper class like:

https://github.com/krasznaa/acts/blob/cuda-improvements-20200826/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/StreamWrapper.hpp

This is not a huge priority to change, I just don't like shared pointers all that much to begin with. (Though I know that Acts uses them extensively...)

paulgessinger replied:
I think the fact that `shared_ptr` works with forward declarations is due to careful implementation. Not arguing that `shared_ptr` should be used here (haven't looked at the code closely), but we generally try to be careful where we use them, and only do so if we feel like we get enough value out of them. Some code has `shared_ptr` for historical reasons, which is going to shrink with #412 for example.

czangela replied:
So I guess the question is, whether we will call the correct destructor to the queue object (which is an incomplete type to the Seedfinder class) when we destruct the Seedfinder object.

Some interesting resources about this:
https://howardhinnant.github.io/incomplete.html
https://stackoverflow.com/questions/2517245/why-really-deleting-an-incomplete-type-is-undefined-behaviour
https://stackoverflow.com/questions/25405135/shared-ptr-destructor-copy-and-incomplete-type

Let's look at an example here: https://wandbox.org/permlink/POjoXf3mt83bbW2k.
I guess I'm doing the same in the real code, but then it means it shouldn't work. (I'm confused, whether I should return the shared_ptr by reference.)

(I was thinking about providing some arguments about the topic, but I ended up so confused, I'd rather say nothing. I think we will call the correct destructor in case our object doesn't get destroyed prematurely.)

I guess we'll end up at rather not using shared_ptr, so I will change the implementation, and use a wrapper object instead (or call an external function from the destructor the deletes the pointer?).

czangela replied:
So I guess the question is, whether we will call the correct destructor to the queue object (which is an incomplete type to the Seedfinder class) when we destruct the Seedfinder object.

Some interesting resources about this:
https://howardhinnant.github.io/incomplete.html
https://stackoverflow.com/questions/2517245/why-really-deleting-an-incomplete-type-is-undefined-behaviour
https://stackoverflow.com/questions/25405135/shared-ptr-destructor-copy-and-incomplete-type

Let's look at an example here: https://wandbox.org/permlink/m0sjZv6jvXmeKtEd.

After all, I think we will call the correct destructor. The shared_ptr works with forward declarations, because all it stores is two pointers, and the "control block" has the destructor information (and other things).

Still, I'm not sure whether we should use this, I can just change the implementation.

(bonus: A [similar topic](https://feherenfekete.wordpress.com/2014/04/24/shared_ptr-polymorphic-magic-pitfall/) I heard a presentation about half a year ago, and I think I understood just now.)

paulgessinger commented:
Thanks for this PR @czangela! Looking forward to mention this, eventually!

paulgessinger commented:
I'm a bit out-of-the-loop on SYCL, but does this now require a special compiler? Or is it just a library? In any case, I think, as with the CUDA plugin, we should add a CI job that builds this plugin, to make sure we don't break it by mistake.

krasznaa commented:
> I'm a bit out-of-the-loop on SYCL, but does this now require a special compiler? Or is it just a library? In any case, I think, as with the CUDA plugin, we should add a CI job that builds this plugin, to make sure we don't break it by mistake.

The way that we were working with this with @czangela, we only developed the code using Intel's [oneAPI](https://software.intel.com/oneAPI) compiler. Actually, we were not even using that, but a hand-built version of their latest development code. You can find some more info about how one can set up such a thing, here: https://gitlab.cern.ch/akraszna/atlas-gpu-devel-env

For a moment I thought that it may be feasible to just use [triSYCL](https://github.com/triSYCL/triSYCL) in the CI instead of Intel's compiler, but I just remembered that triSYCL doesn't actually support USM (Universal Shared Memory). Which, for performance reasons, the plugin code now uses.

So... A Docker image supporting this build could either be based on what I created in the above mentioned repository, or you could also base it directly on the images provided by Intel.

https://hub.docker.com/r/intel/oneapi-basekit

I have to admit I never tried to use their Docker image though...

krasznaa commented:
So... I was trying to build the code this morning, but unfortunately I had to do multiple things to be able to compile the code without any compilation errors or warnings.

```diff
diff --git a/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp b/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp
index 86cf508cc..ff8d1de1f 100644
--- a/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp
+++ b/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp
@@ -23,7 +23,7 @@ class DeviceExperimentCuts {
   /// @param top top space point of the current seed
   /// @return seed weight to be added to the seed's weight
   float seedWeight(const detail::deviceSpacePoint& bottom,
-                   const detail::deviceSpacePoint& middle,
+                   const detail::deviceSpacePoint& /*middle*/,
                    const detail::deviceSpacePoint& top) const {
     float weight = 0;
     if (bottom.r > 150) {
@@ -41,8 +41,8 @@ class DeviceExperimentCuts {
   /// @return true if the seed should be kept, false if the seed should be
   /// discarded
   bool singleSeedCut(float weight, const detail::deviceSpacePoint& bottom,
-                     const detail::deviceSpacePoint& middle,
-                     const detail::deviceSpacePoint& top) const {
+                     const detail::deviceSpacePoint& /*middle*/,
+                     const detail::deviceSpacePoint& /*top*/) const {
     return !(bottom.r > 150. && weight < 380.);
   };
 };
diff --git a/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp b/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp
index 6b78f8f58..f5c530ffd 100644
--- a/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp
+++ b/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp
@@ -6,7 +6,7 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-#include "Acts/Plugins/Sycl/Seeding/Seedfinder.hpp"
+#include "Acts/Plugins/Sycl/Seeding/CreateSeedsForGroupSycl.h"
 #include "Acts/Plugins/Sycl/Seeding/detail/Types.h"
 
 #include <algorithm>
@@ -46,7 +46,7 @@ void createSeedsForGroupSycl(
     const std::vector<detail::deviceSpacePoint>& bottomSPs,
     const std::vector<detail::deviceSpacePoint>& middleSPs,
     const std::vector<detail::deviceSpacePoint>& topSPs,
-    std::vector<std::vector<SeedData>>& seeds) {
+    std::vector<std::vector<detail::SeedData>>& seeds) {
   // Each vector stores data of space points in simplified
   // structures of float variables
   // M: number of middle space points
@@ -74,9 +74,6 @@ void createSeedsForGroupSycl(
   uint32_t edgesComb = 0;
 
   try {
-    using am = cl::sycl::access::mode;
-    using at = cl::sycl::access::target;
-
     uint32_t globalBufferSize =
         q->get_device().get_info<cl::sycl::info::device::global_mem_size>();
     uint32_t maxWorkGroupSize =
@@ -254,10 +251,10 @@ void createSeedsForGroupSycl(
       uint32_t* deviceSumComb = cl::sycl::malloc_device<uint32_t>(M + 1, *q);
 
       // Allocations for coordinate transformation.
-      deviceLinEqCircle* deviceLinBot =
-          cl::sycl::malloc_device<deviceLinEqCircle>(edgesBottom, *q);
-      deviceLinEqCircle* deviceLinTop =
-          cl::sycl::malloc_device<deviceLinEqCircle>(edgesTop, *q);
+      detail::deviceLinEqCircle* deviceLinBot =
+          cl::sycl::malloc_device<detail::deviceLinEqCircle>(edgesBottom, *q);
+      detail::deviceLinEqCircle* deviceLinTop =
+          cl::sycl::malloc_device<detail::deviceLinEqCircle>(edgesTop, *q);
 
       q->memcpy(deviceMidIndPerBot, indMidBotComp.data(),
                 sizeof(uint32_t) * edgesBottom);
@@ -337,7 +334,7 @@ void createSeedsForGroupSycl(
                 const auto iDeltaR = cl::sycl::sqrt(iDeltaR2);
                 const auto cot_theta = -(deltaZ * iDeltaR);
 
-                deviceLinEqCircle L;
+                detail::deviceLinEqCircle L;
                 L.cotTheta = cot_theta;
                 L.zo = zM - rM * cot_theta;
                 L.iDeltaR = iDeltaR;
@@ -380,7 +377,7 @@ void createSeedsForGroupSycl(
                 const auto iDeltaR = cl::sycl::sqrt(iDeltaR2);
                 const auto cot_theta = deltaZ * iDeltaR;
 
-                deviceLinEqCircle L;
+                detail::deviceLinEqCircle L;
                 L.cotTheta = cot_theta;
                 L.zo = zM - rM * cot_theta;
                 L.iDeltaR = iDeltaR;
@@ -406,16 +403,16 @@ void createSeedsForGroupSycl(
       const auto maxMemoryAllocation =
           std::min(edgesComb,
                    globalBufferSize /
-                       uint32_t((sizeof(TripletData) + sizeof(SeedData)) * 2));
+                       uint32_t((sizeof(detail::TripletData) + sizeof(detail::SeedData)) * 2));
 
-      TripletData* deviceCurvImpact =
-          cl::sycl::malloc_device<TripletData>(maxMemoryAllocation, (*q));
+      detail::TripletData* deviceCurvImpact =
+          cl::sycl::malloc_device<detail::TripletData>(maxMemoryAllocation, (*q));
 
       // Reserve memory in advance for seed indices and weight
       // Other way around would allocating it inside the loop
       // -> less memory usage, but more frequent allocation and deallocation
-      SeedData* deviceSeedArray =
-          cl::sycl::malloc_device<SeedData>(maxMemoryAllocation, *q);
+      detail::SeedData* deviceSeedArray =
+          cl::sycl::malloc_device<detail::SeedData>(maxMemoryAllocation, *q);
 
       uint32_t* tripletSearchOffset = cl::sycl::malloc_device<uint32_t>(1, *q);
 
@@ -478,7 +475,7 @@ void createSeedsForGroupSycl(
                   }
                   mid = L;
 
-                  TripletData T = {MIN, MIN};
+                  detail::TripletData T = {MIN, MIN};
                   deviceCurvImpact[idx] = T;
                   const auto numT = deviceNumTopDuplets[mid];
 
@@ -652,7 +649,7 @@ void createSeedsForGroupSycl(
                    if (deviceCuts.singleSeedCut(weight, bottomSP, middleSP,
                                                 topSP)) {
                      const auto i = countSeeds[0].fetch_add(1);
-                     SeedData D;
+                     detail::SeedData D;
                      D.bottom = bot;
                      D.top = top;
                      D.middle = mid;
@@ -668,9 +665,9 @@ void createSeedsForGroupSycl(
             q->memcpy(&sumSeeds, countSeeds, sizeof(std::atomic_uint32_t));
         e0.wait();
 
-        std::vector<SeedData> hostSeedArray(sumSeeds);
+        std::vector<detail::SeedData> hostSeedArray(sumSeeds);
         auto e1 = q->memcpy(&hostSeedArray[0], deviceSeedArray,
-                            sumSeeds * sizeof(SeedData));
+                            sumSeeds * sizeof(detail::SeedData));
         e1.wait();
 
         for (uint32_t t = 0; t < sumSeeds; ++t) {
diff --git a/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp b/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp
index 3bbba9f74..b3acfea18 100644
--- a/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp
+++ b/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp
@@ -13,14 +13,14 @@ struct SpacePoint {
   ;
   SpacePoint(float p_x, float p_y, float p_z, float p_r, int p_surface,
              float p_varianceR, float p_varianceZ, int p_id)
-      : m_id(p_id),
+      : varianceR(p_varianceR),
+        varianceZ(p_varianceZ),
+        m_surface(p_surface),
+        m_id(p_id),
         m_x(p_x),
         m_y(p_y),
         m_z(p_z),
-        m_r(p_r),
-        varianceR(p_varianceR),
-        varianceZ(p_varianceZ),
-        m_surface(p_surface){};
+        m_r(p_r) {}
   float x() const { return m_x; }
   float y() const { return m_y; }
   float z() const { return m_z; }
```

But on the bright side: After compiling the code for my old-ish Intel integrated GPU, I got the following result on it for one of the files that I hacked together yesterday from the ttbar simulation that Noemi gave us:

```
[bash][Elrond]:build > ./bin/ActsUnitTestSeedfinderSycl -f /data/projects/acts/atlas_seeds/pu100/evt1.txt 
Running on: Intel(R) Gen9 HD Graphics NEO
read 186463 SP from file /data/projects/acts/atlas_seeds/pu100/evt1.txt
Preparation time: 0.969350
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

------------------------- Time Metric -------------------------
             Device:        CPU  GPU (SYCL)  Speedup/ Agreement
           Time (s):   3.155872    2.624208            1.202600
---------------------------------------------------------------

[bash][Elrond]:build >
```

(There also seems to be something funky about this printout at the moment btw...)

So, the direction is good, but things definitely still need to change. :wink:

Attila

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/CreateSeedsForGroupSycl.h:
```diff
@@ -0,0 +1,37 @@
+// This file is part of the Acts project.
```
This file will need a `.hpp` extension instead of `.h`.

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/detail/Types.h:
```diff
@@ -0,0 +1,64 @@
+};
+
+// Parameters required to calculate circle with linear equation.
+struct deviceLinEqCircle {
```
Wow... I'm pretty baffled by GitHub right now... Is it not possible to un-resolve a discussion? :confused: In any case... Let's keep the discussion from https://github.com/acts-project/acts/pull/416/files#r481815029 going here then.

The capitalisation of these type names will need to change, according to the Acts coding rules.

czangela replied:
Oh yes, you're right. I was just reading fast, and thought you're talking about starting namespaces with capital letters. Sorry, I'll be more careful next time.

czangela commented:
> So... I was trying to build the code this morning, but unfortunately I had to do multiple things to be able to compile the code without any compilation errors or warnings.
> 
> ```diff
> diff --git a/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp b/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp
> index 86cf508cc..ff8d1de1f 100644
> --- a/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp
> +++ b/Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp
> @@ -23,7 +23,7 @@ class DeviceExperimentCuts {
>    /// @param top top space point of the current seed
>    /// @return seed weight to be added to the seed's weight
>    float seedWeight(const detail::deviceSpacePoint& bottom,
> -                   const detail::deviceSpacePoint& middle,
> +                   const detail::deviceSpacePoint& /*middle*/,
>                     const detail::deviceSpacePoint& top) const {
>      float weight = 0;
>      if (bottom.r > 150) {
> @@ -41,8 +41,8 @@ class DeviceExperimentCuts {
>    /// @return true if the seed should be kept, false if the seed should be
>    /// discarded
>    bool singleSeedCut(float weight, const detail::deviceSpacePoint& bottom,
> -                     const detail::deviceSpacePoint& middle,
> -                     const detail::deviceSpacePoint& top) const {
> +                     const detail::deviceSpacePoint& /*middle*/,
> +                     const detail::deviceSpacePoint& /*top*/) const {
>      return !(bottom.r > 150. && weight < 380.);
>    };
>  };
> diff --git a/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp b/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp
> index 6b78f8f58..f5c530ffd 100644
> --- a/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp
> +++ b/Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp
> @@ -6,7 +6,7 @@
>  // License, v. 2.0. If a copy of the MPL was not distributed with this
>  // file, You can obtain one at http://mozilla.org/MPL/2.0/.
>  
> -#include "Acts/Plugins/Sycl/Seeding/Seedfinder.hpp"
> +#include "Acts/Plugins/Sycl/Seeding/CreateSeedsForGroupSycl.h"
>  #include "Acts/Plugins/Sycl/Seeding/detail/Types.h"
>  
>  #include <algorithm>
> @@ -46,7 +46,7 @@ void createSeedsForGroupSycl(
>      const std::vector<detail::deviceSpacePoint>& bottomSPs,
>      const std::vector<detail::deviceSpacePoint>& middleSPs,
>      const std::vector<detail::deviceSpacePoint>& topSPs,
> -    std::vector<std::vector<SeedData>>& seeds) {
> +    std::vector<std::vector<detail::SeedData>>& seeds) {
>    // Each vector stores data of space points in simplified
>    // structures of float variables
>    // M: number of middle space points
> @@ -74,9 +74,6 @@ void createSeedsForGroupSycl(
>    uint32_t edgesComb = 0;
>  
>    try {
> -    using am = cl::sycl::access::mode;
> -    using at = cl::sycl::access::target;
> -
>      uint32_t globalBufferSize =
>          q->get_device().get_info<cl::sycl::info::device::global_mem_size>();
>      uint32_t maxWorkGroupSize =
> @@ -254,10 +251,10 @@ void createSeedsForGroupSycl(
>        uint32_t* deviceSumComb = cl::sycl::malloc_device<uint32_t>(M + 1, *q);
>  
>        // Allocations for coordinate transformation.
> -      deviceLinEqCircle* deviceLinBot =
> -          cl::sycl::malloc_device<deviceLinEqCircle>(edgesBottom, *q);
> -      deviceLinEqCircle* deviceLinTop =
> -          cl::sycl::malloc_device<deviceLinEqCircle>(edgesTop, *q);
> +      detail::deviceLinEqCircle* deviceLinBot =
> +          cl::sycl::malloc_device<detail::deviceLinEqCircle>(edgesBottom, *q);
> +      detail::deviceLinEqCircle* deviceLinTop =
> +          cl::sycl::malloc_device<detail::deviceLinEqCircle>(edgesTop, *q);
>  
>        q->memcpy(deviceMidIndPerBot, indMidBotComp.data(),
>                  sizeof(uint32_t) * edgesBottom);
> @@ -337,7 +334,7 @@ void createSeedsForGroupSycl(
>                  const auto iDeltaR = cl::sycl::sqrt(iDeltaR2);
>                  const auto cot_theta = -(deltaZ * iDeltaR);
>  
> -                deviceLinEqCircle L;
> +                detail::deviceLinEqCircle L;
>                  L.cotTheta = cot_theta;
>                  L.zo = zM - rM * cot_theta;
>                  L.iDeltaR = iDeltaR;
> @@ -380,7 +377,7 @@ void createSeedsForGroupSycl(
>                  const auto iDeltaR = cl::sycl::sqrt(iDeltaR2);
>                  const auto cot_theta = deltaZ * iDeltaR;
>  
> -                deviceLinEqCircle L;
> +                detail::deviceLinEqCircle L;
>                  L.cotTheta = cot_theta;
>                  L.zo = zM - rM * cot_theta;
>                  L.iDeltaR = iDeltaR;
> @@ -406,16 +403,16 @@ void createSeedsForGroupSycl(
>        const auto maxMemoryAllocation =
>            std::min(edgesComb,
>                     globalBufferSize /
> -                       uint32_t((sizeof(TripletData) + sizeof(SeedData)) * 2));
> +                       uint32_t((sizeof(detail::TripletData) + sizeof(detail::SeedData)) * 2));
>  
> -      TripletData* deviceCurvImpact =
> -          cl::sycl::malloc_device<TripletData>(maxMemoryAllocation, (*q));
> +      detail::TripletData* deviceCurvImpact =
> +          cl::sycl::malloc_device<detail::TripletData>(maxMemoryAllocation, (*q));
>  
>        // Reserve memory in advance for seed indices and weight
>        // Other way around would allocating it inside the loop
>        // -> less memory usage, but more frequent allocation and deallocation
> -      SeedData* deviceSeedArray =
> -          cl::sycl::malloc_device<SeedData>(maxMemoryAllocation, *q);
> +      detail::SeedData* deviceSeedArray =
> +          cl::sycl::malloc_device<detail::SeedData>(maxMemoryAllocation, *q);
>  
>        uint32_t* tripletSearchOffset = cl::sycl::malloc_device<uint32_t>(1, *q);
>  
> @@ -478,7 +475,7 @@ void createSeedsForGroupSycl(
>                    }
>                    mid = L;
>  
> -                  TripletData T = {MIN, MIN};
> +                  detail::TripletData T = {MIN, MIN};
>                    deviceCurvImpact[idx] = T;
>                    const auto numT = deviceNumTopDuplets[mid];
>  
> @@ -652,7 +649,7 @@ void createSeedsForGroupSycl(
>                     if (deviceCuts.singleSeedCut(weight, bottomSP, middleSP,
>                                                  topSP)) {
>                       const auto i = countSeeds[0].fetch_add(1);
> -                     SeedData D;
> +                     detail::SeedData D;
>                       D.bottom = bot;
>                       D.top = top;
>                       D.middle = mid;
> @@ -668,9 +665,9 @@ void createSeedsForGroupSycl(
>              q->memcpy(&sumSeeds, countSeeds, sizeof(std::atomic_uint32_t));
>          e0.wait();
>  
> -        std::vector<SeedData> hostSeedArray(sumSeeds);
> +        std::vector<detail::SeedData> hostSeedArray(sumSeeds);
>          auto e1 = q->memcpy(&hostSeedArray[0], deviceSeedArray,
> -                            sumSeeds * sizeof(SeedData));
> +                            sumSeeds * sizeof(detail::SeedData));
>          e1.wait();
>  
>          for (uint32_t t = 0; t < sumSeeds; ++t) {
> diff --git a/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp b/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp
> index 3bbba9f74..b3acfea18 100644
> --- a/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp
> +++ b/Tests/UnitTests/Plugins/Sycl/Seeding/SpacePoint.hpp
> @@ -13,14 +13,14 @@ struct SpacePoint {
>    ;
>    SpacePoint(float p_x, float p_y, float p_z, float p_r, int p_surface,
>               float p_varianceR, float p_varianceZ, int p_id)
> -      : m_id(p_id),
> +      : varianceR(p_varianceR),
> +        varianceZ(p_varianceZ),
> +        m_surface(p_surface),
> +        m_id(p_id),
>          m_x(p_x),
>          m_y(p_y),
>          m_z(p_z),
> -        m_r(p_r),
> -        varianceR(p_varianceR),
> -        varianceZ(p_varianceZ),
> -        m_surface(p_surface){};
> +        m_r(p_r) {}
>    float x() const { return m_x; }
>    float y() const { return m_y; }
>    float z() const { return m_z; }
> ```
> 
> But on the bright side: After compiling the code for my old-ish Intel integrated GPU, I got the following result on it for one of the files that I hacked together yesterday from the ttbar simulation that Noemi gave us:
> 
> ```
> [bash][Elrond]:build > ./bin/ActsUnitTestSeedfinderSycl -f /data/projects/acts/atlas_seeds/pu100/evt10.txt -m
> Running on: Intel(R) Gen9 HD Graphics NEO
> read 189600 SP from file /data/projects/acts/atlas_seeds/pu100/evt10.txt
> Preparation time: 0.451869
> Analyzed 260 groups for CPU
> Analyzed 260 groups for SYCL
> 
> ------------------------- Time Metric -------------------------
>              Device:        CPU  GPU (SYCL)  Speedup/ Agreement
>            Time (s):   3.419834    2.386087            1.433239
>         Seeds found:      73698       73699           99.971504
> ---------------------------------------------------------------
> 
> [bash][Elrond]:build >
> ```
> 
> So, the direction is good, but things definitely still need to change. 
> 
> Attila

I knew about the warnings, but the code compiles without errors. It was because I left an extra using directive in Seedfinder.ipp, which I shouldn't have. I removed it and added the detail namespace everywhere. Thanks, I overlooked it. :/

krasznaa commented on Plugins/Sycl/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+  include/Acts/Plugins/Sycl/Seeding/Seedfinder.ipp
+  include/Acts/Plugins/Sycl/Utilities/CreateQueue.h
+  include/Acts/Plugins/Sycl/Utilities/DeviceSelector.h
+  include/Acts/Plugins/Sycl/Utilities/ListPlatforms.h
```
As I had to realise, there are a bunch of headers with a `.h` suffix, not just the one that I noticed before. These should all become `.hpp`. (I guess `.h` would be reasonable if it was implemented in a pure C `.c` file, but that's not what we do here.)

Also, don't forget to update the names here, in the CMake configuration. `CreateSeedsForGroupSycl.h` is already renamed to `.hpp`, but the CMake code still has its old name. (Leading to a cmake configuration error.)

krasznaa commented on Plugins/Sycl/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+add_library(
+  ActsPluginSycl SHARED
+  # header files
+  include/Acts/Plugins/Sycl/Seeding/detail/Types.h
```
:confused: Did you leave this as `Types.h` on purpose, or was this just an oversight?

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.ipp:
```diff
@@ -0,0 +1,145 @@
+                                 SP->varianceR(), SP->varianceZ()});
+  }
+
+  const auto deviceConfigData = detail::DeviceSeedfinderConfig{
```
My comment about this variable was also closed without changing the code. :confused: Could you make this a member variable on the class?

czangela replied:
I only looked at the four bottom lines, and thought I know what you're talking about. Turns out I didn't.

czangela replied:
Should be fine now. :)

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.hpp:
```diff
@@ -0,0 +1,68 @@
+
+  Seedfinder(Acts::SeedfinderConfig<external_spacepoint_t> config,
+             const Acts::Sycl::DeviceExperimentCuts& cuts,
+             const Acts::Sycl::QueueWrapper& wrappedQueue);
```
I see your intent here, but I'm not convinced at all that this would result in more readable code than what I had in mind...

Sure, this code can work, but to me it just makes ownership of a queue a lot harder to follow than what I was proposing. And I also didn't see many cases of people using the `&&` operator in anything else than the move constructor or move assignment operator of a class.

Normally you would have just a single constructor to this class.

```c++
  Seedfinder(Acts::SeedfinderConfig<external_spacepoint_t> config,
             const Acts::Sycl::DeviceExperimentCuts& cuts,
             Acts::Sycl::QueueWrapper wrappedQueue);
```

With this, the user could either pass an object to the constructor "by value" (in which case `QueueWrapper`'s copy constructor would be used), or using `std::move(...)` (in which case `QueueWrapper`'s move constructor would be used).

But it still doesn't solve my original issue. :frowning: What if the client code already has a `cl::sysl::queue` object that they want the seedfinder to use? How would they pass **that** in?

czangela replied:
I didn't know that, but I'll change the constructors. :)

I was thinking a lot about the original issue. I think it is better to manage (creating and destroying) queue-s through a class.

Our intention with hiding sycl headers in the first place, was that we can hide it's implementation details. With expecting a pointer to the queue, we expect that the other TU see's sycl headers (they would definitely need them for destroying the queue object).

(Also this way, we don't have to forward declare the type every time we use it.)

I know it's not clear where the ownership is in this version, but I think it's logical. If you create a new object, you're the owner. If you move, you're also owner. If you make a copy, than you don't own anything.
This also allows dereferencing null pointer, in case you make a copy and the copy doesn't know about the pointer's death. Still, the other version you proposed would have this issue as well.

krasznaa replied:
> I would also suggest that you move your example code from `Tests/UnitTests/Plugins/Sycl` to `Tests/IntegrationTests/Sycl`. Unit tests should be automated tests using the chosen unit test framework, which your example is not. Integration tests are not as tightly defined and better match your use case.

I think this is a good idea. Let me just point out that @czangela didn't come up with this location by herself. She was copying what I did, and what @beomki-yeo did before me. :stuck_out_tongue:
  - https://github.com/acts-project/acts/tree/master/Tests/UnitTests/Plugins/Cuda/Seeding
  - https://github.com/acts-project/acts/tree/master/Tests/UnitTests/Plugins/Cuda/Seeding2

And we just copied what you did for the CPU implementation. :stuck_out_tongue:

https://github.com/acts-project/acts/tree/master/Tests/UnitTests/Core/Seeding

So yes, if we want to move the executable to the integration tests, let's do that. We will just need to remember to move all the other executables later on as well.

krasznaa replied:
I have no idea why GitHub put this message here. :confused:

czangela commented:
A few things about the implementation:

### 1. Memory management

#### 1.1. Duplet search and coordinate transformation

I decided to store data about duplet pairs (indices), and corresponding data of their linear equation form with no offset.
That would mean for example, if I have 3 compatible bottom space points for the first middle space point, 2 for the second, 0 and 5 for the third and fourth, then I could store them like this:

|sp0_e0|sp0_e1|sp0_e2|sp1_e0|sp1_e1|sp3_e0|sp3_e1|sp3_e2|sp3_e3|sp3_e4|
|---|---|---|---|---|---|---|---|---|---|

Another option would be to determine a maximum for each middle space point, which would be 5 in this case, and reserve that much space for each of them. This causes us to reserve more space and start more threads in kernels than necessary. Also, we would need an extra condition inside the kernel, to decide whether the thread we're executing on corresponds to a good duplet pair.

I found this to more of an overhead than storing data sequentially with no waste and retrieving middle space point indices in a bit trickier way (which makes computations more expensive).

(It could be that I'm wrong, also I did not store performance metrics about this, so I cannot show anything.)

During the duplet search, I count compatible pairs, and I make prefix sum arrays for the counted duplets (with a leading zero) **for indexing**.
```cpp
  // Up to the Nth space point, the sum of the compatible bottom/top space points.
  // We need these for indexing other vectors later in the algorithm.
  // These are prefix sum arrays, with a leading zero.
  std::vector<uint32_t> sumBotCompUptoMid(M + 1, 0);
  std::vector<uint32_t> sumTopCompUptoMid(M + 1, 0);
  std::vector<uint32_t> sumBotTopCombined(M + 1, 0);
```

#### 1.2. Triplet search

I agree, that this is a bit ugly.


So we can't fit all the triplet combinations into memory at once (at least for larger test files), so we have to do something else.
I made a version previously, where I start kernels for each middle space point separately. (This was really slow for small test files, because of the overhead of starting a kernel.)

There was also a version, which reserved constant space (=let it be C) for triplets per middle space point, and started kernels equal to globalMemorySize / C. (This seemed like a good idea, but it meant starting not so much fewer kernels but reserving considerably more memory (comparing to the first approach). It was due to that in the first version we could omit starting kernels for middle space points, that had either no compatible bottom or top space point.)

(What I didn't try, but could be good, is skipping middle space points with no compatible pairs, and not reserving memory and threads for them. That would certainly do better than the second version, but I'm not sure it would do better than the current version.)

There were others, that don't deserve to be mentioned here.

**What I'm doing now**, is storing triplets sequentially (again, I'm trying not to send more data and start more threads than necessary), and calculating middle space point indices with a binomial search inside a kernel (yes, this is expensive but less so than other approaches).

I have a few (very professional) images (with nice colors) that may help me out here.

<img src="https://user-images.githubusercontent.com/33939088/92103755-b1d1df00-ede0-11ea-8870-9cb3dddb8ba1.png" alt="duplet_pairs" width="500">

(Green corresponds to bottom, purple to middle and blue to top space points.)

<img src="https://user-images.githubusercontent.com/33939088/92104514-b64ac780-ede1-11ea-9239-74793a82ceaf.png" alt="triplet search" width="400">

So I'm always counting memory requirements, and starting kernels when we meet a limit.

```cpp
// Determine the interval [firstMiddle, lastMiddle) right end based
// on memory requirements.
while (lastMiddle + 1 <= M &&
      (sumBotTopCombined[lastMiddle + 1] - sumBotTopCombined[firstMiddle] < maxMemoryAllocation)) {
          ++lastMiddle;
}
```

czangela commented:
### 2. Experiment specific cuts

We're doing experiment specific cuts inside the kernel (filter for 2 space points fixed -> bottom and middle).

```cpp
weight += deviceCuts.seedWeight(bottomSP, middleSP, topSP);

if (deviceCuts.singleSeedCut(weight, bottomSP, middleSP, topSP)) {
  ...
}
```

Inside the kernel we're calling member functions of the Acts::Sycl::DeviceExperimentCuts class, which are also implemented in the header files. They could be moved to separate translation units if they were marked with the SYCL_EXTERNAL attribute, and we included the CL/sycl.hpp header files which would require linking sycl to Acts Core. We're not doing that right now.

Another option would be to use dynamic inheritance, make the functions virtual, and have them implemented by the user. SYCL currently doesn't allow virtual functions inside kernels. (More specifically, it only takes objects as parameters, which are trivially copyable, and classes with virtual functions are not.)

Or we could use function pointers, which I'm not sure are supported right now. (This is what @krasznaa is doing with CUDA.)

There might as well be other options I (and @krasznaa) haven't thought about.

This implementation is meant to be *temporary*, as it would require the user to rewrite the DeviceExperimentCuts.hpp header file and implement their own experiment specific cuts.

Currently, ATLAS cuts are included.

```cpp
float seedWeight(const detail::DeviceSpacePoint& bottom,
                 const detail::DeviceSpacePoint& /*middle*/,
                 const detail::DeviceSpacePoint& top) const {
  float weight = 0;
  if (bottom.r > 150.f) {
    weight = 400.f;
  }
  if (top.r < 150.f) {
    weight = 200.f;
  }
  return weight;
};

bool singleSeedCut(float weight, const detail::DeviceSpacePoint& bottom,
                   const detail::DeviceSpacePoint& /*middle*/,
                   const detail::DeviceSpacePoint& /*top*/) const {
  return !(bottom.r > 150.f && weight < 380.f);
};
```

czangela commented:
### 3. Running the code
We can list available platforms and devices with **clinfo**:
```bash
[bash][atspot01]:build > clinfo
Number of platforms                               3
  Platform Name                                   Intel(R) OpenCL
  Platform Vendor                                 Intel(R) Corporation
  Platform Version                                OpenCL 2.1 LINUX
...
```
This is useful to decide which device we want to generate binaries for because we need to give that to CMake.

#### 3.1. Building with CMake
First of all, the ACTS_BUILD_PLUGIN_SYCL option should be switched on.
By default, we generate binaries for SPIR 64 targets.

```cmake
# Figure out which SYCL target platforms are available.
set( SYCL_POSSIBLE_TARGETS "spir64-unknown-unknown-sycldevice"
...
```
We can set the SYCL_POSSIBLE_TARGETS variable something else, such as "nvptx64-nvidia-cuda-sycldevice" for CUDA.

#### 3.2. Trying the code

We can build a given unit test (which is everything but not a unit test), for this purpose.
So we have ActsUnitTestSeedfinderSycl, which takes a bunch of command-line options.
```
Allowed options:
  -h [ --help ]           Print usage message.
  -f [ --FILE ] arg       Provide path for input file.
  -n [ --NUM ] arg (=500) Number of groups to iterate in seed finding.
  -d [ --DEVICE ] arg     Provide a substring of the preferred device.
  -l [ --LIST ]           List available SYCL platforms and devices.
  -G [ --GPU ]            Execute code only on gpu. Default is 0.
  -a [ --ALL ]            Analyze all groups. Default is 0.
  -m [ --MATCH ]          Count seed matches. Default is 0.
```

We try to select the preferred device by the -d option, or we could also set the **SYCL_BE** environment variable, which makes SYCL runtime recognize only the chosen devices. Such as SYCL_BE=PI_OPENCL. (Btw. BE stands for backend.)

The input file should have a predefined format:
```bash
[bash][atspot01]:build > head ../../seeds/sp1.txt 
lxyz 2 53.8945 1.9094 -1502.5 0.05 0.1
lxyz 2 99.428 14.8217 -1502.5 0.05 0.1
lxyz 2 38.0541 3.87263 -1502.5 0.05 0.1
lxyz 2 74.6186 0.982534 -1502.5 0.05 0.1
lxyz 2 32.8574 2.5107 -1502.5 0.05 0.1
```
Numbers correspond to layer, x, y, z coordinates, R, and Z variances (?).

In the end we should see something like this:
```bash
[bash][atspot01]:build > bin/ActsUnitTestSeedfinderSycl -f ../../seeds/sp1.txt -m -d GeForce
Running on: GeForce RTX 2060
read 85808 SP from file ../../seeds/sp1.txt
Preparation time: 0.694717
Analyzed 260 groups for CPU
Analyzed 260 groups for SYCL

--------------------------- Time Metric ---------------------------
             Device:          CPU   GPU (SYCL)  Speedup/ Agreement
           Time (s):     0.623309     0.278775            2.235889
        Seeds found:        19759        19759          100.000000
-------------------------------------------------------------------
```

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.ipp:
```diff
@@ -0,0 +1,135 @@
+    : m_config(config),
+      m_deviceCuts(cuts),
+      m_wrappedQueue(std::move(wrappedQueue)) {
+  init();
```
I'm giving the code one last read-through.

At this point I believe the code is in a reasonable shape to be merged in. I just want to point out a couple of things that I think could help you in your work later on.

I would not have used a separate `init()` function here. As that's a very [C\+\+03](https://en.wikipedia.org/wiki/C%2B%2B03) thing to do. :stuck_out_tongue: I would rather have done one of the following:
  - Have just one constructor, that receives a default-constructed `Acts::Sycl::QueueWrapper` object as its last argument. Which the client may or may not override.
  - Have 2 constructors defined. Put the full initialisation of the object into the constructor with the larger number of arguments, and have the constructor with fewer arguments call the other constructor. Like:

```c++
template <typename external_spacepoint_t>
Seedfinder<external_spacepoint_t>::Seedfinder(
    Acts::SeedfinderConfig<external_spacepoint_t> config,
    const Acts::Sycl::DeviceExperimentCuts& cuts)
    : Seedfinder(config, cuts, Acts::Sycl::QueueWrapper()) {}
```

czangela replied:
I wanted to use delegating constructors at first, but this didn't come to my mind. Neither did the first version. But I like that very much. I'll change this soon.

krasznaa commented on Plugins/Sycl/src/Utilities/QueueWrapper.cpp:
```diff
@@ -0,0 +1,96 @@
+
+QueueWrapper& QueueWrapper::operator=(const QueueWrapper& other) {
+  m_queue = other.m_queue;
+  m_ownsQueue = false;
```
This might actually be good to fix still...

You must put a `this == &other` check into this operator as well. Otherwise self-assignments would behave in a **very** broken way. (Leaking memory...)

czangela replied:
Can I resolve this?

czangela replied:
Very true.

krasznaa commented on Tests/UnitTests/Plugins/Sycl/Seeding/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+add_executable(ActsUnitTestSeedfinderSycl SeedfinderSyclTest.cpp CommandLineArguments.cpp)
+FIND_PACKAGE(Boost COMPONENTS program_options REQUIRED)
+INCLUDE_DIRECTORIES(${Boost_INCLUDE_DIRS})
+TARGET_LINK_LIBRARIES(ActsUnitTestSeedfinderSycl PRIVATE ${unittest_extra_libraries} ${Boost_LIBRARIES})
```
Ideally these 4 lines should rather just look like:

```cmake
add_executable(ActsUnitTestSeedfinderSycl
  SeedfinderSyclTest.cpp CommandLineArguments.h CommandLineArguments.cpp
  AtlasCuts.hpp SpacePoint.hpp)
target_link_libraries(ActsUnitTestSeedfinderSycl
  PRIVATE Boost::program_options ActsCore ActsPluginSycl)
```

I'm not a big fan of using `${unittest_extra_libraries}`, but my bigger point is that you should use the `Boost::program_options` imported target. Which is available to this configuration file even without calling `find_package(...)` in it explicitly.

msmk0 replied:
The `${unittest_extra_libraries}` are only needed if you are using the custom `add_unittest` macro to add the executables. You can do without here.

As mentioned above, the `find_package` calls are listed in a central location and should not be called again here.

krasznaa commented on Tests/UnitTests/Plugins/Sycl/Seeding/CommandLineArguments.cpp:
```diff
@@ -0,0 +1,61 @@
+
+namespace Acts {
+namespace Sycl {
+extern void listPlatforms();
```
Wait... What's this? Why didn't you just include the appropriate header file from the plugin? Using `extern` here seems unnecessary anyway... :confused:

krasznaa commented:
Okay, that's all from me... :wink:

paulgessinger commented:
> Okay, that's all from me... 😉

@krasznaa, if you're happy with this including the changes required, can you add your approval via the interface?

msmk0 commented on Plugins/Sycl/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+# This module needs at least CMake 3.13.
+cmake_minimum_required( VERSION 3.13 )
```
I do not think this is needed since it is already part of the `FindSYCL` module.

msmk0 commented on Plugins/Sycl/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+# This module needs at least CMake 3.13.
+cmake_minimum_required( VERSION 3.13 )
+
+find_package(SYCL REQUIRED)
```
Could you move the `find_package` call to the main `CMakeLists.txt` in the repository root. Within Acts the `find_package` calls are all listed at a single point and not distributed over many files so it is easier to see what are the requirements.

czangela replied:
I hope I did an OK job here.

msmk0 commented on Tests/UnitTests/Plugins/CMakeLists.txt:
```diff
@@ -2,3 +2,4 @@ add_subdirectory_if(Digitization ACTS_BUILD_PLUGIN_DIGITIZATION)
 add_subdirectory_if(Json ACTS_BUILD_PLUGIN_JSON)
 add_subdirectory_if(Cuda ACTS_BUILD_PLUGIN_CUDA)
 add_subdirectory_if(TGeo ACTS_BUILD_PLUGIN_TGEO)
+add_subdirectory_if(Sycl ACTS_BUILD_PLUGIN_SYCL)
```
```suggestion
add_subdirectory_if(Sycl ACTS_BUILD_PLUGIN_SYCL)
add_subdirectory_if(TGeo ACTS_BUILD_PLUGIN_TGEO)
```
Alphabetical order.

msmk0 commented on cmake/FindSYCL.cmake:
```diff
@@ -0,0 +1,193 @@
+#
+# Variable(s) set up by the module:
+#  - SYCL_FOUND: Boolean, set to TRUE when SYCL compilation is available.
+#  - acts_setup_sycl_target( <target> [DEPENDENCY PUBLIC|PRIVATE|INTERFACE] ):
```
As far as I understand it, this CMake module is not Acts specific? For consistency with other CMake command, the helper function should probably be called `target_setup_sycl` or `acts_target_setup_sycl` if you want to retain the `acts_` prefix. This matches e.g. `target_link_libraries(...)`.

krasznaa replied:
Hi Moritz,

The naming here is consistent with the naming that I chose in the ATLAS offline code.

https://gitlab.cern.ch/atlas/atlasexternals/-/blob/master/Build/AtlasCMake/modules/FindSYCL.cmake

I see your point, but it's not really a strong one. This function is not all that analogous to [target_link_libraries(...)](https://cmake.org/cmake/help/latest/command/target_link_libraries.html) that we should use that as a template for its name.

Since we were the ones who wrote this code, the `acts_` prefix seems appropriate to me. Intel did not produce any convenient CMake integration for their compiler just yet...

Cheers,
        Attila

czangela replied:
I chose the later option. :)

(p.s. I hope you don't think I was able to write this CMake file.)

msmk0 replied:
@krasznaa Thanks for the clarification. My line of though was that it modifies an existing target similar to what most of the `target_...` functions do. I am Ok either way.

msmk0 reviewed (approved):
@czangela Thank you for this nice addition. I have not comments on the code since @krasznaa was already doing an excellent job reviewing. I only looked at the CMake integration and added a few suggestions so that this plugins integrates well with the rest of the Acts codebase.

I would also suggest that you move your example code from `Tests/UnitTests/Plugins/Sycl` to `Tests/IntegrationTests/Sycl`. Unit tests should be automated tests using the chosen unit test framework, which your example is not. Integration tests are not as tightly defined and better match your use case.

krasznaa commented:
> > Okay, that's all from me... wink
> 
> @krasznaa, once you're happy with this including the changes required, can you add your approval via the interface?

I don't believe that I can do that. I think that's only allowed for people with elevated privileges on the repository.

Or I can just not figure out where this would be in the UI... That's also possible.

krasznaa commented on Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp:
```diff
@@ -0,0 +1,316 @@
+// This file is part of the Acts project.
```
Okay, still learning about the GitHub UI... I guess this is how I would give my approval...

paulgessinger replied:
Yes! You don't even need to add any additional comments.

krasznaa reviewed (approved):
For me, the PR is in a good enough shape at this point.

paulgessinger commented:
@krasznaa, hm, I think anyone can leave a review, but it won't automatically fulfill our 'review required' check if you're not on approver list. I might be wrong, but if I'm right, there should be an "Review changes" button on the top right after you click on the "Files changed" tab.

I was more thinking that this is a good way for you to indicate that you're ok to merge this. Otherwise you can also do this as a comment or so.

czangela commented on CMakeLists.txt:
```diff
@@ -33,6 +33,7 @@ option(ACTS_BUILD_PLUGIN_JSON "Build json plugin" OFF)
 option(ACTS_USE_SYSTEM_NLOHMANN_JSON "Use nlohmann::json provided by the system instead of the bundled version" OFF)
 option(ACTS_BUILD_PLUGIN_LEGACY "Build legacy plugin" OFF)
 option(ACTS_BUILD_PLUGIN_TGEO "Build TGeo plugin" OFF)
+option(ACTS_BUILD_PLUGIN_SYCL "Build SYCL plugin" OFF)
```
It makes sense now. :D

czangela commented on Plugins/CMakeLists.txt:
```diff
@@ -9,5 +9,6 @@ add_component_if(Legacy PluginLegacy ACTS_BUILD_PLUGIN_LEGACY)
 # dependent plugins. depend either on a independent plugins or on one another
 add_component_if(TGeo PluginTGeo ACTS_BUILD_PLUGIN_TGEO)
 add_component_if(DD4hep PluginDD4hep ACTS_BUILD_PLUGIN_DD4HEP)
+add_component_if(Sycl PluginSycl ACTS_BUILD_PLUGIN_SYCL)
```
Yes, thank you.

paulgessinger commented:
@czangela would you mind putting the doc-style comments that you added in this PR into our docs as well? That way it's more easily preserved, and we don't have to go back to this PR to find it.

krasznaa commented:
> @czangela would you mind putting the doc-style comments that you added in this PR into our docs as well? That way it's more easily preserved, and we don't have to go back to this PR to find it. (This can be done in a separate PR)

Since her summer studentship is finishing at the end of today, I was planning to do that myself next week. Along with some documentation for the CUDA seedfinder code.

msmk0 commented:
It should be fine, if you add the documentation in a separate PR.

czangela commented:
> > @czangela would you mind putting the doc-style comments that you added in this PR into our docs as well? That way it's more easily preserved, and we don't have to go back to this PR to find it. (This can be done in a separate PR)
> 
> Since her summer studentship is finishing at the end of today, I was planning to do that myself next week. Along with some documentation for the CUDA seedfinder code.

Sure, I'll do that. :) I just wanted to leave them in the comments as well.

robertlangenberg commented on Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp:
```diff
@@ -0,0 +1,719 @@
+
+      q->submit([&](cl::sycl::handler& h) {
+        h.parallel_for<duplet_search_top_kernel>(
+            cl::sycl::range<2>{M, B}, [=](cl::sycl::id<2> idx) {
```
Just to mark the bug you had found yourself
```suggestion
            cl::sycl::range<2>{M, T}, [=](cl::sycl::id<2> idx) {
```

czangela replied:
It wasn't really me but thanks,

robertlangenberg reviewed (changes requested):
As discussed, if you could add a short comment what is happening when calculating indices, that would go a long way to figure out down the road what's happening. Thanks!

robertlangenberg commented on Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp:
```diff
@@ -0,0 +1,719 @@
+          cl::sycl::malloc_device<std::atomic_uint32_t>(1, *q);
+
+      seeds.resize(M);
+      const float MIN = -100000.f;
```
I didn't understand how this exactly works, did I understand correctly you could replace it with a bool?
To give you an intuition for the range: the helix curvature is actually stored as 1/diameter, so it's going to be between -1 and 1 (actually much smaller as we usually want helix radii >500mm). 
The maximum allowed impact parameters are configurable and are always >=0

czangela replied:
I tried that, but I guess too much padding is added to the struct this way, and the extra memory allocations make the code slower.

robertlangenberg replied:
Then never mind, should be a safe value (-1 would also have been small enough). Good to go in from my side!

robertlangenberg reviewed (approved):
This looks very good, thank you Angela and Attila!
